internal/app/machined/pkg/controllers/config/node_labels.go

```diff
@@ -0,0 +1,436 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package config
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/go-pointer"
+	"go.uber.org/zap"
+
+	"github.com/talos-systems/talos/pkg/kubernetes"
+	talosconfig "github.com/talos-systems/talos/pkg/machinery/config"
+	"github.com/talos-systems/talos/pkg/machinery/config/types/v1alpha1/machine"
+	"github.com/talos-systems/talos/pkg/machinery/resources/config"
+	"github.com/talos-systems/talos/pkg/machinery/resources/k8s"
+	"github.com/talos-systems/talos/pkg/machinery/resources/secrets"
+)
+
+// NodeLabelsK8sClientIface is the minimal interface needed fo manage the node labels.
```

Comment:
can I get the tl;dr on what is the design of this controller?

if the labels owned by Talos are kept as annotations in the Kubernetes `Node` resource, it should be fairly straightforward to sync whatever labels present in the machine config with the `Node` resource?

Reply:
The controller first synchronizes the labels from the machine config and from kubernetes to the COSI state. I wanted to have the labels from other sources also in Talos so the user can see which labels are already "occupied". Here I had to take care that labels which are removed on k8s from another source get removed in the Talos state aswell.
In a second step all the labels from the COSI state are synchronized to k8s if they are owned by Talos. I made sure that no labels that aren't owned by Talos are overwritten or removed. Here I decided to add and remove each label in an individual patch, so that if there is a permission error for one of the labels, all the other labels still get applied.

Reply:
do we actually need to go that far? 

we could present all node labels, by 'Watch'ing the `Node` resource, but in the end Talos should only remove/update labels which are recorded as managed by Talos in the annotation. 

so there's definitely 'Status' part of that we could do, that is the list of labels present today

there's also 'Spec' part which syncs labels from the (machine) config to the node

I'm also thinking if we should split 'NodeLabel' spec into separate resources? this would allow to inject e.g. controlplane label based on machine type.

Reply:
makes sense, I split them up into NodeLabelStatus and NodeLabelSpec resources. The code should be less convoluted now.